caom2pipe/data_source_composable.py

```diff
@@ -175,6 +189,130 @@ def get_work(self):
+        self._logger.debug('End get_work')
+        return self._work
+
+    def _append_work(self, entry):
+        with os.scandir(entry) as dir_listing:
+            for entry in dir_listing:
+                if entry.is_dir() and self._recursive:
+                    self._append_work(entry.path)
+                else:
+                    for extension in self._extensions:
+                        if entry.name.endswith(extension):
+                            self._logger.debug(
+                                f'Adding {entry.path} to work list.'
+                            )
+                            self._work.append(entry.path)
+                            break
+
+
+class ListDirTimeBoxDataSource(DataSource):
+    """
+    A time-boxed directory listing of all .fits* and .hdf5 files. The time-box
+    is based on os.stat.st_mtime for a file.
+
+    Implementation choices:
+    - one or many directories? one is fire-and-forget, many is probably
+      more accurate of operational conditions - e.g. VLASS1.1, VLASS1.2,
+      etc, OR vos:Nat/NIFS/...., OR NEOSSAT's NESS, 2017, 2018, 2019, etc
+    - send it off to glob, walk, or do it semi-here? one/a few high-level
+      directories that point to a lot of occupied storage space could take a
+      very long time, and a lot of memory, in glob/walk
+    """
+
+    def __init__(self, config, recursive=True):
+        """
+
+        :param config: manage_composable.Config
+        :param recursive: True if sub-directories should also be checked
+        """
+        super(ListDirTimeBoxDataSource, self).__init__(config)
+        self._source_directories = config.data_sources
+        self._extensions = config.data_source_extensions
+        self._recursive = recursive
+        self._work = deque()
+        self._temp = defaultdict(list)
+        self._logger = logging.getLogger(self.__class__.__name__)
+
+    def get_time_box_work(self, prev_exec_time, exec_time):
+        """
+        :param prev_exec_time datetime start of the timestamp chunk
+        :param exec_time datetime end of the timestamp chunk
+        :return: a deque of StateRunnerMeta instances, with
+            prev_exec_time <= os.stat.mtime <= exec_time, and sorted by
+            os.stat.mtime
+        """
+        self._logger.debug(
+            f'Begin get_time_box_work from {prev_exec_time} to {exec_time}.'
+        )
+        for source in self._source_directories:
+            self._logger.debug(f'Looking for work in {source}')
+            self._append_work(prev_exec_time, exec_time, source)
+        # ensure the result returned is sorted by timestamp in ascending
+        # order
+        for mtime in sorted(self._temp):
+            for entry in self._temp[mtime]:
+                self._work.append(
+                    StateRunnerMeta(entry_name=entry, entry_ts=mtime)
+                )
+        self._logger.debug('End get_time_box_work')
+        self._temp = defaultdict(list)
```

Comment:
Why dictionary here? You might be able to achieve the same thing with the `filter` function (make sure it file and dates with the interval) and `sorted` but I don't know how performant that solution is.

Reply:
I chose a structure that keeps both the timestamp, and the file name, because the `sort` is on the file modification time, so that if the processing of a time-box chunk is interrupted, it's possible to re-start the processing at a point where only new files will be encountered.

This is going to run at CFHT, on their VM, so at this point I have no performance metrics at all to judge whether or not this is a good decision.